src/native/eventpipe/ep-buffer-manager.c

```diff
@@ -1209,12 +1213,33 @@ ep_buffer_manager_write_all_buffers_to_file_v4 (
 					// miscategorize it, but that seems unlikely.
 					uint32_t last_read_delta = last_read_sequence_number - thread_sequence_number;
 					if (0 < last_read_delta && last_read_delta < 0x80000000) {
-						ep_rt_thread_sequence_number_map_add_or_replace (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), session_state, last_read_sequence_number);
-						if (!exists)
+						if (exists) {
+							ep_rt_thread_sequence_number_map_remove (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), session_state);
+						} else {
 							ep_thread_addref (ep_thread_holder_get_thread (ep_thread_session_state_get_thread_holder_ref (session_state)));
+						}
+						ep_rt_thread_sequence_number_map_add (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), session_state, last_read_sequence_number);
+					}
+
+					// if a session_state was exhausted during this sequence point, mark it for deletion
+					if (ep_thread_session_state_get_buffer_list(session_state)->head_buffer == NULL) {
+
+						if (ep_rt_volatile_load_uint32_t_without_barrier(&(ep_thread_session_state_get_thread(session_state))->unregistered) > 0) {
+
+							ep_rt_thread_session_state_array_append(&session_states_to_delete, session_state);
+						}
 					}
 					ep_rt_thread_session_state_list_iterator_next (&thread_session_state_list_iterator);
 				}
+
+				// foreach session_state_to_delete in session_states_to_delete
+				for (ep_rt_thread_session_state_array_iterator_t thread_session_state_array_iterator = ep_rt_thread_session_state_array_iterator_begin(&session_states_to_delete);
+					 !ep_rt_thread_session_state_array_iterator_end(&session_states_to_delete, &thread_session_state_array_iterator);
+					 ep_rt_thread_session_state_array_iterator_next(&thread_session_state_array_iterator)) {
+
+					EventPipeThreadSessionState * session_state = ep_rt_thread_session_state_array_iterator_value(&thread_session_state_array_iterator);
+					ep_rt_thread_session_state_list_remove(&buffer_manager->thread_session_state_list, session_state);
```

Comment:
Maybe it could be worth adding support to remove an item in list based on iterator and return a new iterator as result (similar to stl::list::erase, then there wouldn't be a need for this second loop while still holding lock.

Reply:
I'd need to go take a look at the implementation of both Mono's list and Coreclr's SList. I agree though.  I think in the long run we should implement a set of container classes that are shared between the two runtimes or at least used in shared native code like this.

Reply:
For list I believe it could be done since iterator just needs to move to a position in the list that continue to work after the item as been removed. I believe we should be able to implement it on both list implementation and be able to drop this second loop.